waku/node/delivery_monitor/send_monitor.nim

```diff
@@ -0,0 +1,212 @@
+    wakuLightpushClient.addPublishObserver(sendMonitor)
+
+  return ok(sendMonitor)
+
+proc performFeedbackAndCleanup(
+    self: SendMonitor,
+    msgsToDiscard: Table[WakuMessageHash, DeliveryInfo],
+    success: DeliverySuccess,
+    dir: DeliveryDirection,
+    comment: string,
+) =
+  ## This procs allows to bring delivery feedback to the API client
+  ## It requires a 'deliveryCb' to be registered beforehand.
+  if self.deliveryCb.isNil():
+    error "deliveryCb is nil in performFeedbackAndCleanup",
+      success, dir, comment, hashes = toSeq(msgsToDiscard.keys).mapIt(shortLog(it))
+    return
+
+  for hash, deliveryInfo in msgsToDiscard:
+    debug "send monitor performFeedbackAndCleanup",
+      success, dir, comment, msg_hash = shortLog(hash)
+
+    self.deliveryCb(success, dir, comment, hash, deliveryInfo.msg)
+    self.publishedMessages.del(hash)
+
+proc checkMsgsInStore(
+    self: SendMonitor, msgsToValidate: Table[WakuMessageHash, DeliveryInfo]
+): Future[
+    Result[
+      tuple[
+        publishedCorrectly: Table[WakuMessageHash, DeliveryInfo],
+        notYetPublished: Table[WakuMessageHash, DeliveryInfo],
+      ],
+      void,
+    ]
+] {.async.} =
+  let hashesToValidate = toSeq(msgsToValidate.keys)
+
+  let storeResp: StoreQueryResponse = (
+    await self.storeClient.queryToAny(
+      StoreQueryRequest(includeData: false, messageHashes: hashesToValidate)
+    )
+  ).valueOr:
+    error "checkMsgsInStore failed to get remote msgHashes",
+      hashes = hashesToValidate.mapIt(shortLog(it)), error = $error
+    return err()
+
+  let publishedHashes = storeResp.messages.mapIt(it.messageHash)
+
+  var notYetPublished: Table[WakuMessageHash, DeliveryInfo]
+  var publishedCorrectly: Table[WakuMessageHash, DeliveryInfo]
+
+  for msgHash, deliveryInfo in msgsToValidate.pairs:
+    if publishedHashes.contains(msgHash):
+      publishedCorrectly[msgHash] = deliveryInfo
+      self.publishedMessages.del(msgHash) ## we will no longer track that message
+    else:
+      notYetPublished[msgHash] = deliveryInfo
+
+  return ok((publishedCorrectly: publishedCorrectly, notYetPublished: notYetPublished))
+
+proc processMessages(self: SendMonitor) {.async.} =
+  var msgsToValidate: Table[WakuMessageHash, DeliveryInfo]
+  var msgsToDiscard: Table[WakuMessageHash, DeliveryInfo]
+
+  let now = getNowInNanosecondTime()
+  let timeToCheckThreshold = now - ArchiveTime.nanos
+  let maxLifeTime = now - MaxTimeInCache.nanos
+
+  for hash, deliveryInfo in self.publishedMessages.pairs:
+    if deliveryInfo.msg.timestamp < maxLifeTime:
+      ## message is too old
+      msgsToDiscard[hash] = deliveryInfo
+
+    if deliveryInfo.msg.timestamp < timeToCheckThreshold:
+      msgsToValidate[hash] = deliveryInfo
+
+  ## Discard the messages that are too old
+  self.performFeedbackAndCleanup(
+    msgsToDiscard, DeliverySuccess.UNSUCCESSFUL, DeliveryDirection.PUBLISHING,
+    "Could not publish messages. Please try again.",
```

Comment:
> Maybe should we detail that they weren't published because the timestamps were too old?

Well, in that particular case, the messages' timestamps are not the cause of not being published properly. I'd say that this is just a consequence of not being published properly after a few attempts :) Let me know if that makes sense

Reply:
Sorry! Not sure I understand.

In https://github.com/waku-org/nwaku/blob/7b372b0b09257dbfdc120233a03d77a59c57123d/waku/node/delivery_monitor/send_monitor.nim#L144-L146

Aren't we adding the messages to `msgsToDiscard` because of being too old?

Reply:
> Aren't we adding the messages to `msgsToDiscard` because of being too old?

hahah yes exactly but this is a consequence of not being sent correctly in previous attempts. I think we are saying the same :) 

If we give the response "The timestamp was too old" that can make the user think that the timestamp set in the passed message was set too old already.
Anyways, we can revisit that in the future when we start using it in _Status_.

Thanks for the comments!